tikv-client/src/test/java/com/pingcap/tikv/key/IndexKeyTest.java

```diff
@@ -48,6 +58,87 @@ public void createTest() {
     }
   }
 
+  @Test
+  public void encodeIndexDataValuesTest() {
+    TiColumnInfo col1 = new TiColumnInfo(1, "a", 0, IntegerType.INT, false);
+    TiColumnInfo col2 = new TiColumnInfo(1, "a", 1, StringType.VARCHAR, true);
+    List<TiColumnInfo> tableColumns = new ArrayList<>();
+    tableColumns.add(col1);
+    tableColumns.add(col2);
+    TiTableInfo tableInfo =
+        new TiTableInfo(
+            1,
+            CIStr.newCIStr("test"),
+            "",
+            "",
+            false,
+            true,
+            tableColumns,
+            null,
+            "",
+            0,
+            0,
+            0,
+            0,
+            null,
+            null,
+            null,
+            1,
+            1,
+            0,
+            null,
+            0);
+
+    // CommonHandle
+    TiIndexColumn index1 = new TiIndexColumn(CIStr.newCIStr("a"), 0, DataType.UNSPECIFIED_LEN);
+    List<TiIndexColumn> indexColumns1 = new ArrayList<>();
+    indexColumns1.add(index1);
+    Object[][] testRows1 =
+        new Object[][] {
+          new Object[] {null, "1"}, new Object[] {2, "2"},
+        };
+    Key[][] expectations1 =
+        new Key[][] {
+          new Key[] {
+            TypedKey.toTypedKey(null, IntegerType.INT), TypedKey.toTypedKey("1", StringType.VARCHAR)
+          },
+          new Key[] {TypedKey.toTypedKey(2, IntegerType.INT)},
+        };
+    for (int i = 0; i < testRows1.length; i++) {
+      Row row = ObjectRowImpl.create(testRows1[i]);
+      Handle handle =
+          CommonHandle.newCommonHandle(
+              new DataType[] {StringType.VARCHAR}, new Object[] {row.get(1, StringType.VARCHAR)});
+      IndexKey.EncodeIndexDataResult result =
+          IndexKey.encodeIndexDataValues(row, indexColumns1, handle, true, tableInfo);
+      assertArrayEquals(expectations1[i], result.keys);
+    }
+
+    // PkHandle
+    TiIndexColumn index2 = new TiIndexColumn(CIStr.newCIStr("b"), 1, DataType.UNSPECIFIED_LEN);
+    List<TiIndexColumn> indexColumns2 = new ArrayList<>();
+    Object[][] testRows2 =
+        new Object[][] {
+          new Object[] {1, null}, new Object[] {2, "2"},
+        };
+    Key[][] expectations2 =
+        new Key[][] {
+          new Key[] {
+            TypedKey.toTypedKey(null, StringType.VARCHAR),
+            TypedKey.toTypedKey(1, IntegerType.BIGINT)
+          },
+          new Key[] {TypedKey.toTypedKey("2", StringType.VARCHAR)},
+        };
+    indexColumns2.add(index2);
+    for (int i = 0; i < testRows2.length; i++) {
+      Row row = ObjectRowImpl.create(testRows2[i]);
+      Handle handle = new IntHandle((Integer) row.get(0, IntegerType.BIGINT));
+      IndexKey.EncodeIndexDataResult result =
+          IndexKey.encodeIndexDataValues(row, indexColumns2, handle, true, tableInfo);
+      assertArrayEquals(expectations2[i], result.keys);
```

Comment:
we need not alter tableInfo's pkHandle to true?

Reply:
> we need not alter tableInfo's pkHandle to true?

we need. I changed